snuba/query/processors/mapping_promoter.py

```diff
@@ -0,0 +1,142 @@
+
+    val_column, val_field = value_col_split
+    key_column, key_field = key_col_split
+
+    if val_column != key_column:
+        return None
+
+    key_literal = exp.parameters[1].parameters[1]
+    if not isinstance(key_literal.value, str):
+        return None
+
+    return SubscriptableMatch(
+        table_name=exp.parameters[0].table_name,
+        column_name=val_column,
+        key_field=key_field,
+        val_field=val_field,
+        key=key_literal.value,
+    )
+
+
+class PromotedColumnsSpec(NamedTuple):
+    # The name of the key field in the mapping column
+    key_field: str
+    # The name of the value field in the mapping column
+    val_field: str
+    # The mapping between keys in the column and real columns
+    column_mapping: Mapping[str, str]
+
+
+class MappingColumnPromoter(QueryProcessor):
+    """
+    Promotes expressions that access the value of a mapping column by
+    replacing them with the corresponding promoted column provided in
+    the constructor.
+
+    Example: tags["myTag"]
+        -> arrayElement("tags.value", indexOf("tags.key", "myTag"))
+        -> toString(promoted_MyTag)
+
+    This happens if there is a promoted_MyTag column in the storage
+    that maps to the myTag tag.
+    """
+
+    def __init__(
+        self, columns: ColumnSet, mapping_spec: Mapping[str, PromotedColumnsSpec]
+    ) -> None:
+        # The ColumnSet of the dataset. Used to format promoted
+        # columns with the right type.
+        self.__columns = columns
+        # The configuration for this processor. The key of the
+        # mapping is the name of the nested column.
+        self.__spec = mapping_spec
+
+    def process_query(self, query: Query, request_settings: RequestSettings) -> None:
+        def transform_nested_column(exp: Expression) -> Expression:
+            subscript = match_subscriptable_reference(exp)
+            if subscript is None:
+                return exp
+
+            if (
+                subscript.column_name in self.__spec
+                and subscript.val_field == self.__spec[subscript.column_name].val_field
+                and subscript.key_field == self.__spec[subscript.column_name].key_field
+            ):
+                promoted_col_name = self.__spec[
+                    subscript.column_name
+                ].column_mapping.get(subscript.key)
+                if promoted_col_name:
+                    col_type = self.__columns.get(promoted_col_name, None)
+                    col_type_name = str(col_type) if col_type else None
+
+                    ret_col = Column(exp.alias, promoted_col_name, subscript.table_name)
+                    if (
+                        col_type_name
+                        and "String" in col_type_name
+                        and "FixedString" not in col_type_name
+                    ):
+                        return ret_col
+                    else:
+                        return FunctionCall(exp.alias, "toString", (ret_col,))
```

Comment:
This is a copy of what the current tag promotion logic does. I think it is not covering all cases.

Reply:
A comment here would be really helpful explaining this behavior (not what it is doing, but why it is doing it.)

Reply:
ok